bee-bundle/src/transaction.rs

```diff
@@ -3,197 +3,119 @@ use crate::constants::{ADDRESS, BUNDLE_HASH, IOTA_SUPPLY, NONCE, PAYLOAD, TAG};
 use bee_common::constants::{TRANSACTION_TRYT_LEN, TRYTE_ZERO};
 use bee_common::Errors;
 use bee_common::Result as BeeResult;
-use bee_common::Tryte;
-use bee_ternary::IsTryte;
+use bee_ternary::{util::trytes_to_trits_buf, IsTryte, T1B1Buf, TritBuf};
 
 use std::fmt;
-use std::hash::Hash as StdHash;
-use std::hash::Hasher as StdHasher;
-
-macro_rules! implement_debug {
-    ($($t:ty),+) => {
-    $(
-        impl fmt::Debug for $t {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                write!(f, "{}", self.0.iter().collect::<String>())
-            }
-        }
-    )+
-    }
-}
-
-macro_rules! implement_display {
-    ($($t:ty),+) => {
-    $(
-        impl fmt::Display for $t {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                write!(f, "{}", self.0.iter().collect::<String>())
-            }
-        }
-    )+
-    }
-}
-
-macro_rules! implement_hash {
-    ($($t:ty),+) => {
-    $(
-        impl StdHash for $t {
-            fn hash<H : StdHasher>(&self, state: &mut H) {
-                    self.0.hash(state);
-            }
-        }
-    )+
-    }
-}
-
-macro_rules! implement_eq {
-    ($($t:ty),+) => {
-    $(
-        impl PartialEq for $t {
-            fn eq(&self,other: &$t) -> bool {
-                    self.0.iter().zip(other.0.iter()).all(|(a,b)| a == b)
-            }
-        }
-
-        impl Eq for $t {}
-
-    )+
-    }
-}
-
-macro_rules! implement_clone {
-    ($($t:ty),+) => {
-    $(
-        impl Clone for $t {
-            fn clone(&self) -> Self {
-                    Self(self.0)
-            }
-        }
 
-    )+
-    }
-}
-
-pub struct Payload(pub [Tryte; PAYLOAD.tryte_offset.length]);
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Payload(TritBuf<T1B1Buf>);
 
 impl Payload {
     pub fn zeros() -> Self {
-        Self([TRYTE_ZERO; PAYLOAD.tryte_offset.length])
+        Self(TritBuf::zeros(PAYLOAD.trit_offset.length))
     }
 
     pub fn from_str(payload: &str) -> Self {
         assert!(payload.len() <= PAYLOAD.tryte_offset.length);
         assert!(payload.chars().all(|c| c.is_tryte()));
 
-        let mut trytes = [TRYTE_ZERO; PAYLOAD.tryte_offset.length];
+        let missing_trytes = PAYLOAD.tryte_offset.length - payload.len();
+        let payload = payload.to_owned() + &TRYTE_ZERO.to_string().repeat(missing_trytes);
```

Comment:
please use map collect instead

Reply:
scrap what I said in my previous comment.
best solution is to use something like this:
`std::iter::repeat(TRYTE_ZERO).take(missing_trytes).collect::<String>();`

another very minor thing (mostly a matter of taste I guess):
I would change the variable name 'missing_trytes' to 'num_missing_trytes' as this the information we are really storing, and not trytes themselves. You know what I mean?